src/imars3d/backend/workflow/engine.py

```diff
@@ -0,0 +1,208 @@
+# package imports
+from imars3d.backend.workflow import validate
+
+# third party imports
+from param import ParameterizedFunction
+
+# standard imports
+from collections import namedtuple
+from collections.abc import Iterable
+import importlib
+from typing import Any, List, Optional, Union
+
+
+class WorkflowEngineError(RuntimeError):
+    pass
+
+
+def WorkflowEngine(config: Union[str, dict], schema: str = validate.SCHEMA) -> "WorkflowEngineWithSchema":  # noqa F821
+    r"""engine that interprets and runs all the operations in the reconstruction configuration
+
+    Parameters
+    ----------
+    config
+        JSON configuration for reconstruction-reduction
+    schema
+        JSON schema for validation of `config`
+
+    Returns
+    -------
+        Instance of class WorkflowEngineWithSchema
+    """
+
+    class WorkflowEngineWithSchema:
```

Comment:
What is the reason for using an embedded class here?
Why not making it a module level class and make the function call it?